File: frontend/trader/src/lib/api/insurance.ts
/**
 * Trade Insurance API client.
 * Backed by /api/v1/insurance/* — see backend/services/gateway/src/api/insurance.py
 */
import api from './client';

// Backend returns the admin-defined tier label verbatim. Legacy mode
// returns one of basic/advanced/pro/elite; simple mode returns whatever
// admin typed (e.g. "50%", "70%"). UI must render it as-is — never key
// off a fixed enum.
export type InsuranceTier = string;

export interface TierQuote {
  tier: InsuranceTier;
  fee: number;
  coverage_pct: number;
  max_cap: number;
  estimated_refund: number;
  risk_score: number;
}

export interface QuoteRequest {
  account_id: string;
  symbol: string;
  side: 'buy' | 'sell';
  lots: number;
  leverage?: number;
  stop_loss?: number;
  take_profit?: number;
}

export interface ActivateResponse {
  policy_id: string;
  fee_charged: string;
  status: 'active';
}

export interface PolicyOut {
  id: string;
  position_id: string | null;
  instrument_symbol: string | null;
  tier: InsuranceTier;
  fee: string;
  coverage_pct: string;
  max_cap: string;
  status: 'active' | 'claimed' | 'expired' | 'denied';
  activated_at: string;
  settled_at: string | null;
  // Why the policy ended up denied/expired. Backend persists a short
  // code (min_duration, daily_claim_limit, not_a_loss, …); the UI maps
  // it to a human-readable explanation.
  settled_reason: string | null;
}

export type ClaimStatus = 'pending' | 'paid';

export interface ClaimOut {
  id: string;
  policy_id: string;
  loss_amount: string;
  claim_amount: string;
  status: ClaimStatus;
  paid_at: string | null;
  claimed_at: string | null;
  instrument_symbol: string | null;
  tier: InsuranceTier | null;
}

export interface ClaimPayResponse {
  claim_id: string;
  amount: string;
  credited_to: 'credit' | 'balance';
  status: 'paid';
}

export const insuranceApi = {
  quote: (body: QuoteRequest) => api.post<TierQuote[]>('/insurance/quote', body),
  activate: (position_id: string, tier: InsuranceTier) =>
    api.post<ActivateResponse>('/insurance/activate', { position_id, tier }),
  active: () => api.get<PolicyOut[]>('/insurance/active'),
  policies: (limit = 50) => api.get<PolicyOut[]>(`/insurance/policies?limit=${limit}`),
  claims: (limit = 50, status?: ClaimStatus) => {
    const params = new URLSearchParams({ limit: String(limit) });
    if (status) params.set('status', status);
    return api.get<ClaimOut[]>(`/insurance/claims?${params.toString()}`);
  },
  claimPayout: (claimId: string) =>
    api.post<ClaimPayResponse>(`/insurance/claims/${claimId}/claim`, {}),
};

File: backend/packages/common/src/schemas/insurance.py
"""Trade Insurance API Pydantic schemas."""
from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class InsuranceQuoteRequest(BaseModel):
    account_id: UUID
    symbol: str
    side: str = Field(..., pattern="^(buy|sell)$")
    lots: Decimal = Field(gt=0, le=100)
    leverage: int = Field(default=100, ge=1, le=2000)
    stop_loss: Optional[Decimal] = None
    take_profit: Optional[Decimal] = None


class InsuranceTierQuote(BaseModel):
    tier: str
    fee: float
    coverage_pct: float
    max_cap: float
    estimated_refund: float
    risk_score: float


class InsuranceActivateRequest(BaseModel):
    # Tier is the admin-defined label string (e.g. "50%", "70%"). The
    # legacy 4-tier regex (basic/advanced/pro/elite) was dropped during
    # the 2026-05-25 cleanup — any non-empty label up to 32 chars is OK.
    position_id: UUID
    tier: str = Field(..., min_length=1, max_length=32)


class InsuranceActivateResponse(BaseModel):
    policy_id: UUID
    fee_charged: Decimal
    status: str


class InsurancePolicyOut(BaseModel):
    id: UUID
    position_id: Optional[UUID] = None
    instrument_symbol: Optional[str] = None
    tier: str
    fee: Decimal
    coverage_pct: Decimal
    max_cap: Decimal
    status: str
    activated_at: datetime
    settled_at: Optional[datetime] = None
    # Why the policy ended up denied/expired (min_duration,
    # daily_claim_limit, not_a_loss, …). NULL while active or
    # successfully claimed. Frontend maps the code to a friendly label.
    settled_reason: Optional[str] = None


class InsuranceClaimOut(BaseModel):
    id: UUID
    policy_id: UUID
    loss_amount: Decimal
    claim_amount: Decimal
    # 'pending' = trader hasn't pressed Claim yet, 'paid' = credited.
    status: str
    # Pending claims have no paid_at / claimed_at yet.
    paid_at: Optional[datetime] = None
    claimed_at: Optional[datetime] = None
    instrument_symbol: Optional[str] = None
    tier: Optional[str] = None


class InsuranceClaimPayResponse(BaseModel):
    claim_id: UUID
    amount: Decimal
    credited_to: str  # 'credit' or 'balance' (depends on payout_to_credit cfg)
    status: str       # 'paid'
